strawberryfields/backends/tfbackend/ops.py

```diff
@@ -389,6 +395,106 @@ def two_mode_squeezer_matrix(theta, phi, cutoff, batched=False, dtype=tf.complex
     )
 
 
+def choi_trick(S, d, dtype=tf.complex64):
+    """Transforms the parameter from S,d to R, y, C corresponding to the parameters of the multidimensional hermite polynomials"""
+    m = S.shape[0] // 2
+    S = tf.cast(S, dtype=dtype)
+    d = tf.cast(d, dtype=dtype)
+    # m: num of modes
+    choi_r = tf.cast(tf.math.asinh(1.0), dtype=dtype)
+    ch = tf.math.cosh(choi_r) * tf.eye(m, dtype=dtype)
+    sh = tf.math.sinh(choi_r) * tf.eye(m, dtype=dtype)
+    zh = tf.zeros([m, m], dtype=dtype)
+    Schoi = tf.concat(
+        [
+            tf.concat([ch, sh, zh, zh], 0),
+            tf.concat([sh, ch, zh, zh], 0),
+            tf.concat([zh, zh, ch, -sh], 0),
+            tf.concat([zh, zh, -sh, ch], 0),
+        ],
+        1,
+    )
+    Sxx = S[:m, :m]
+    Sxp = S[:m, m:]
+    Spx = S[m:, :m]
+    Spp = S[m:, m:]
+    idl = tf.eye(m, dtype=dtype)
+    S_exp = (
+        tf.concat(
+            [
+                tf.concat([Sxx, zh, Sxp, zh], 1),
+                tf.concat([zh, idl, zh, zh], 1),
+                tf.concat([Spx, zh, Spp, zh], 1),
+                tf.concat([zh, zh, zh, idl], 1),
+            ],
+            0,
+        )
+        @ Schoi
+    )
+    choi_cov = 0.5 * S_exp @ tf.transpose(S_exp)
+    idl = tf.eye(2 * m, dtype=dtype)
+    Rot = tf.cast(tf.math.sqrt(0.5), dtype=dtype) * tf.concat(
+        [tf.concat([idl, 1j * idl], 1), tf.concat([idl, -1j * idl], 1)], 0
+    )
+    sigma = Rot @ choi_cov @ tf.transpose(tf.math.conj(Rot))
+    zh = tf.zeros([2 * m, 2 * m], dtype=dtype)
+    X = tf.concat([tf.concat([zh, idl], 1), tf.concat([idl, zh], 1)], 0)
+    sigma_Q = sigma + 0.5 * tf.eye(4 * m, dtype=dtype)
+    A_mat = X @ (tf.eye(4 * m, dtype=dtype) - tf.linalg.inv(sigma_Q))
+    E = tf.linalg.diag(
+        tf.concat([tf.ones([m], dtype=dtype), tf.ones([m], dtype=dtype) / tf.math.tanh(choi_r)], 0)
+    )
+    R = -tf.math.conj(E @ A_mat[: 2 * m, : 2 * m] @ E)
+    y = tf.concat(
+        [
+            tf.linalg.matvec(R[:m, :m], tf.math.conj(d)) + tf.transpose(d),
+            tf.linalg.matvec(R[m:, :m], tf.math.conj(d)),
+        ],
+        0,
+    )
+    alpha = tf.concat([tf.cast(d, dtype=dtype), tf.zeros(m, dtype=dtype)], 0)
+    zeta = alpha + tf.linalg.matvec(tf.cast(R, dtype=dtype), tf.math.conj(alpha))
+    C = tf.math.sqrt(
+        tf.math.sqrt(tf.linalg.det(tf.eye(m, dtype=dtype) - R[:m, :m] @ tf.math.conj(R[:m, :m])))
+    ) * tf.exp(-0.5 * tf.reduce_sum(tf.math.conj(alpha) * zeta))
+    return R, y, C
+
+
+@tf.custom_gradient
+def single_gaussian_gate_matrix(R, y, C, cutoff, dtype=tf.complex64.as_numpy_dtype):
+    """creates a N-mode gaussian gate matrix"""
+    C = C.numpy()
+    y = y.numpy()
+    R = R.numpy()
+    gate = gaussian_gate_tw(R, cutoff, y, C=C, dtype=dtype)
+
+    @tf.function
+    def grad(dy):
```

Comment:
I know `dy` is kind of standard, but it would probably be helpful for future maintenance to call this variable with the da_db syntax (in this case what is it? dL_dG?)

Reply:
oh, it's the conjugate of dL_dG